src/core/dash_test.cc

```diff
@@ -365,13 +371,40 @@ TEST_F(DashTest, BumpUp) {
   EXPECT_EQ(1, segment_.CVCOnBump(2, kSecondStashId, 9, hash, touched_bid));
   EXPECT_EQ(touched_bid[0], kSecondStashId);
 
-  segment_.BumpUp(kSecondStashId, 9, hash);
+  segment_.BumpUp(kSecondStashId, 9, hash, RelaxedBumpPolicy{});
   ASSERT_TRUE(key == segment_.Key(0, kNumSlots - 1) || key == segment_.Key(1, kNumSlots - 1));
   EXPECT_TRUE(segment_.GetBucket(kSecondStashId).IsFull());
   EXPECT_TRUE(Contains(key));
   EXPECT_TRUE(segment_.Key(kSecondStashId, 9));
 }
 
+TEST_F(DashTest, BumpPolicy) {
+  struct RestrictedBumpPolicy {
+    bool CanBumpDown(uint64_t key) const {
+      return false;
+    }
+  };
+
+  set<Segment::Key_t> keys = FillSegment(0);
+  constexpr unsigned kFirstStashId = Segment::kNumBuckets;
+
+  EXPECT_TRUE(segment_.GetBucket(0).IsFull());
+  EXPECT_TRUE(segment_.GetBucket(1).IsFull());
+  EXPECT_TRUE(segment_.GetBucket(kFirstStashId).IsFull());
+
+  // check items are immovable in bucket
+  Segment::Key_t key = segment_.Key(1, 2);
+  uint64_t hash = dt_.DoHash(key);
+  segment_.BumpUp(1, 2, hash, RestrictedBumpPolicy{});
+  EXPECT_EQ(key, segment_.Key(1, 2));
+
+  // check items don't swap from stash
+  key = segment_.Key(kFirstStashId, 2);
+  hash = dt_.DoHash(key);
+  segment_.BumpUp(kFirstStashId, 2, hash, RestrictedBumpPolicy{});
+  EXPECT_EQ(key, segment_.Key(kFirstStashId, 2));
```

Comment:
We could check here that the base buckets stash info didn't change. This would require making some more accessors public, like at least `BucketIndex` to find out what the base bucket is before call.